QBO: Make it ~possible~ easier to test in a local dev environment

coreymckrill opened the pull request:
This adds a `local` value for the `WORDCAMP_ENVIRONMENT` constant, and makes various places in the code behave differently when this value is set. This makes it much easier to test interactions with the QBO API and between our own QBO client/server setup.

**To test**

This is specifically designed to work in the Docker environment, but you can probably make it work in other local environments if you change your `WORDCAMP_ENVIRONMENT` constant value to `local` (and maybe update other values in your **wp-config.php** file, see the changeset)

1. After you check out this branch, you may need to rebuild your Docker container to make sure it has the right values in **wp-config.php**.
1. Make sure you have a **sandbox-functionality.php** file in the **mu-plugins** directory.
1. Log in to [Intuit's developer site](https://developer.intuit.com) and go to the My Apps tab. "WordCamp Quickbooks" is the app we want. Go to the Keys & OAuth tab under Development.
1. Add this snippet to your **sandbox-functionality.php** file, using the Client ID and Client Secret from the Intuit app:
    ```php
	// QBO
	add_filter( 'wordcamp_qbo_client_config', function( $config = array() ) {
		$config = array_merge(
			$config,
			array(
				'ClientID'     => 'paste value here',
				'ClientSecret' => 'paste value here',
				'baseUrl'      => 'Development',
			)
		);
	
		return $config;
	}, 99 );
    ```
1. In another browser tab, go to the [Quickbooks settings page](https://wordcamp.test/wp-admin/network/settings.php?page=quickbooks) in the test WordCamp network admin. Click _Connect_ and go through the authentication flow. Choose "Sandbox WPCS" for the company. At the end, you should end up back on the Quickbooks settings page, and it should say "Connected to Sandbox WPCS. Expires in 3 months."
1. Now open a third browser tab and go to one of the other test WordCamp sites. Create a new sponsor invoice, and submit it.
1. Back on the browser tab with the network admin, go to the [Sponsor Invoices Dashboard](https://wordcamp.test/wp-admin/network/admin.php?page=sponsor-invoices-dashboard). You should see the invoice you just created.
1. Click the _Approve_ button on the invoice. This sends the invoice to QBO. You should get a green success notice.
1. Back on the Intuit developer browser tab, under the API Docs & Tools tab, click Sandbox. Then choose the Sandbox WPCS company. This will take you to a QBO dashboard.
1. Click the Invoices item under the Sales tab. You should see the invoice you just approved.

coreymckrill committed:
Docker: Change `WORDCAMP_ENVIRONMENT` value to "local"

This differentiates it from the value on sandboxes.

coreymckrill committed:
QBO Client: Change redirect URI depending on the current environment

Our app on QBO can have multiple valid redirect URIs defined. We have
two: one for production (and thus also our sandboxes), and one for
local testing. This automates the process of switching the URI and
avoids accidentally committing the test URI to production.

coreymckrill committed:
Add helper function for determining the current environment

coreymckrill committed:
QBO Client: Change api base value depending on the environment

coreymckrill committed:
QBO Client: Disable `sslverify` for local environment requests

Our local environment uses self-signed ssl certs, which do not pass
a verification test, which causes our remote post requests to fail in
some cases.

coreymckrill committed:
QBO Client: Show more informative message for unexpected errors

coreymckrill committed:
Docker: Add info about sandbox-functionality.php

coreymckrill committed:
Docker: Add dummy values to QBO constants in wp-config

ryelle commented:
About to start testing this, but it looks like `sandbox-functionality.php` is not ignored in git.

coreymckrill commented:
Oh weird, I didn't even think of that. It looks like I added it to **.git/info/exclude** at some point.

I'll add it to the real .gitignore file.

coreymckrill committed:
Add sandbox-functionality.php to gitignore

ryelle commented on .docker/wp-config.php:
```diff
@@ -1,6 +1,6 @@
 <?php
 
-define( 'WORDCAMP_ENVIRONMENT', 'development' );
+define( 'WORDCAMP_ENVIRONMENT', 'local' );
```
Will this affect other things using `WORDCAMP_ENVIRONMENT`? A quick look seems to show that most conditionals are checking for `production` and fall back to dev settings in all other cases, so it might be fine.

coreymckrill replied:
Yeah, as far as I can tell there isn't any code that is specifically looking for `development` in that constant.

iandunn replied:
After digging through the codebase, it's starting to feel like there could be a lot of side-effects and added complexity around this. Or it's also possible that we're making different assumptions, and some more explicit code might help clarify that.

What problems did you run into while testing locally that prompted this? I don't remember ever having any, so maybe we've been doing it different ways?

What's the reason for introducing a distinction between local and remote sandboxes?

iandunn replied:
Here are the specific things I think might be good to change, if my assumptions are correct. Or at the very least maybe this will help reveal where we might have different assumptions.

`SHOULD` - If we add `local`, I think we should also change the remote sandbox value to be `staging` or `remote`. `development` seems like it'd be easily confused in that context, since conventionally that means a local environment in my experience.

- [ ] `MUST`: It seems like the `wcorg_sponsor_payment_stripe` and `wordcamp_google_maps_api_key` closures in `wcorg-misc.php` would need to be updated, otherwise the keys won't be populated because there's no `local` case in the `switch`.
- [ ] `MUST`: `wcpt-event/notification.php` should probably return early on local sandboxes instead of remote.

`MIGHT` - Would it be simpler to add `case 'local'` to `set_client_config()` in `quickbooks.php`, instead of devs needing to manually add a new `wordcamp_qbo_client_config` callback in `sandbox-functionality.php`?

`MIGHT` - Right now `set_client_config()` is using the sandbox keys in local and remote sandboxes, but it might make more sense to use the production keys in `staging`/`remote` and the sandbox keys in `local`.

Related in Trello: `/c/rErUVQ6P` (first 2 uncompleted checklist items) and `/c/5exJNacO`

coreymckrill replied:
> What problems did you run into while testing locally that prompted this?

The problem was that there were several places in the code that had to be temporarily modified in order to be able to test workflows involving QBO (hardcoded API URLs, the SSL verification setting, etc). This meant finding and changing things in several different files, leaving the working directory dirty, and introducing the risk of accidentally committing these temporary changes.

> What's the reason for introducing a distinction between local and remote sandboxes?

In a local dev environment, the domain is **wordcamp.test**. In both production and on sandboxes, the domain is **wordcamp.org**. There are other differences as well, such as that sandboxes can have two-way communication with other remote servers (e.g. Jetpack connection stuff), while local environments can't (unless you use ngrok or something)

coreymckrill replied:
> If we add local, I think we should also change the remote sandbox value to be staging or remote.

Ok, although I don't think that needs to be part of this PR, since this is specifically about making it easier to test QBO locally.

> It seems like the wcorg_sponsor_payment_stripe and wordcamp_google_maps_api_key closures in wcorg-misc.php would need to be updated, otherwise the keys won't be populated because there's no local case in the switch.

In both of those closures, the `default` switch case catches "local", so it gets the same treatment as if it was "development".

> wcpt-event/notification.php should probably return early on local sandboxes instead of remote.

Good call. I guess we could just change the line at the top to

```php
if ( 'local' === WORDCAMP_ENVIRONMENT ) {
	return;
}
```

> Would it be simpler to add case 'local' to set_client_config() in quickbooks.php, instead of devs needing to manually add a new wordcamp_qbo_client_config callback in sandbox-functionality.php?

We don't want to publish our QBO sandbox keys in the repo, so the values would need to go in **sandbox-functionality.php** either way. Otherwise you end up putting those values in a file under version control and risk accidentally committing them.

> Right now set_client_config() is using the sandbox keys in local and remote sandboxes, but it might make more sense to use the production keys in staging/remote and the sandbox keys in local.

I think remote sandboxes should still use sandbox keys. If we're testing interactions with the QBO API on a sandbox, we wouldn't want test data to get mixed in with actual transactions.

coreymckrill replied:
Re: this comment in Trello:

> The various usages of `WORDCAMP_ENVIRONMENT` throughout the codebase are starting to feel kind of `WET`, because they're not assuming that it exists. It does (and should/must) exist in all of our environments, though, so that feels like it's adding a lot of needless complexity. What about just removing all `defined()` calls and fallbacks etc, and just referencing the value directly everywhere?

The `get_wordcamp_environment()` function introduced in this PR is meant to address this. I didn't go through and replace all the references to the constant in the whole codebase, because that seemed out of scope, but we could easily do that in a separate PR once this is merged.

coreymckrill replied:
> wcpt-event/notification.php should probably return early on local sandboxes instead of remote.

1bb2e2eca0ba693df20964dd844375e344652835

iandunn replied:
> I don't think [changing `development` to `staging`] needs to be part of this PR

I didn't realize you were seeing environments being [_both_ `local` and `development` simultaneously](
https://github.com/WordPress/wordcamp.org/pull/371#discussion_r381501047), rather than changing `development` to _only_ mean remote sandboxes. So I guess in that case it shouldn't change 👍 



> In both of those closures, the default switch case catches "local", so it gets the same treatment as if it was "development".

Ah, you're right, I missed that. 👍 


> just change the line at the top to...

👍 


> the values would need to go in sandbox-functionality.php either way. Otherwise [...] version control

Ah, I didn't realize that your `public_html/wp-config.php` gets overwritten when you re-provision. 👍 


> I think remote sandboxes should still use sandbox keys. If we're testing interactions with the QBO API on a sandbox, we wouldn't want test data to get mixed in with actual transactions.

I guess we just use remote sandboxes differently. I only use mine as a staging environment, like trying to reproduce a bug someone reported w/ production data. Either way works 👍

coreymckrill committed:
Add note about WORDCAMP_QBO_* values

ryelle reviewed (approved):
After some setup woes, it worked as expected and I was even able to debug my own issues 🙂 

I would feel better if @iandunn goes through it too, since he's got a better grasp on all the edge cases.

iandunn commented on .docker/wp-config.php:
```diff
@@ -103,13 +103,14 @@
-define( 'WORDCAMP_SANDBOX_QBO_CLIENT_ID',        '' );
-define( 'WORDCAMP_SANDBOX_QBO_CLIENT_SECRET',    '' );
+define( 'WORDCAMP_SANDBOX_QBO_CLIENT_ID',        "(Optional) your app's development client ID key goes here" );
+define( 'WORDCAMP_SANDBOX_QBO_CLIENT_SECRET',    "(Optional) your app's development client secret key goes here" );
```
Do the apps have separate dev/prod secrets? I only see the production values on the QBO site.

coreymckrill replied:
They do. Here's where to find them:

![qbo-keys](https://user-images.githubusercontent.com/916023/74868768-a4a14700-530b-11ea-8395-2823e67e8426.jpg)

iandunn replied:
Doh, you're right, I totally missed that.

iandunn commented on .docker/wp-config.php:
```diff
@@ -103,13 +103,14 @@
-define( 'WORDCAMP_SANDBOX_QBO_CLIENT_ID',        '' );
-define( 'WORDCAMP_SANDBOX_QBO_CLIENT_SECRET',    '' );
+define( 'WORDCAMP_SANDBOX_QBO_CLIENT_ID',        "(Optional) your app's development client ID key goes here" );
+define( 'WORDCAMP_SANDBOX_QBO_CLIENT_SECRET',    "(Optional) your app's development client secret key goes here" );
```
> Add this snippet to your sandbox-functionality.php file, using the Client ID and Client Secret from the [production] Intuit app:

What's the reason to switch to using the production app in local environments, instead of the sandbox app (even if  sandbox companies are used in both cases) ?

How are you weighing the security implications of having production keys in local environments?

coreymckrill replied:
I guess the testing instructions are a little convoluted since it's spread between steps 3 and 4. But in Step 3 I'm saying to go to the _Development_ keys tab, not _Production_.

So I'm not suggesting that we use production keys in a local environment.

iandunn replied:
👍

iandunn commented on public_html/wp-content/mu-plugins/wcorg-misc.php:
```diff
@@ -4,6 +4,22 @@
+	}
+
+	return $environment;
+}
```
What's the reason for falling back to `dev` here instead of `local` ? Same for `set_client_config()`.

`MIGHT` - `3-helpers-misc.php` seems like a more consistent location for this type of thing.

coreymckrill replied:
> What's the reason for falling back to dev here instead of local ? Same for set_client_config().

I was thinking that `development` is the catchall non-production value, whereas `local` has specific implications since the domain is different, no 2-way communication, etc.

> 3-helpers-misc.php seems like a more consistent location for this type of thing.

Oh, good call. It's kind of confusing that we have **3-helpers-misc.php** _and_ **wcorg-misc.php**. But that's a conversation for another day 😄

coreymckrill replied:
1228241190643ce36608c8fee54750db47b68cca

iandunn replied:
Yeah, I can see that. I don't think the two names were intended to imply a relationship between the two files, just to distinguish miscellaneous helpers from `wcpt`-specific helpers.

Feel free to rename it to whatever makes more sense to you (now or in the future).

iandunn replied:
> I was thinking that `development` is the catchall non-production value, whereas `local` has specific implications since the domain is different, no 2-way communication, etc.

That feels counter-intuitive and unconventional to me. If you'd prefer that, I think all the possible values of `WORDCAMP_ENVIRONMENT` should be documented alongside the `wp-config` definition.

coreymckrill replied:
Documented: 6f7dccda

iandunn commented on public_html/wp-content/plugins/wordcamp-qbo-client/wordcamp-qbo-client.php:
```diff
@@ -302,7 +307,7 @@ public static function send_invoice_to_quickbooks( $invoice_id ) {
 				$sent = $body->message;
 			} else {
-				$sent = 'Unknown error.';
+				$sent = print_r( array( $body ), true );
```
Would `$body` ever contain any sensitive values?

coreymckrill replied:
Not in my testing, but I suppose it's possible it could contain a key or something. I'll rethink the approach here.

coreymckrill replied:
2111916a

coreymckrill committed:
Simplify environment check for Slack notifier

coreymckrill committed:
Move get_wordcamp_environment() to 3-helpers-misc.php

coreymckrill committed:
QBO: Revert change to error handling in `send_invoice_to_quickbooks`

`print_r`-ing the body of the response could potentially leak
sensitive information like keys.

coreymckrill committed:
Document possible values of `WORDCAMP_ENVIRONMENT`
